src/nvim/eval.c

```diff
@@ -13358,26 +13358,38 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
     rettv_dict_alloc(rettv);
     if (rhs != NULL) {
       // Return a dictionary.
-      char_u *lhs = str2special_save(mp->m_keys, true);
-      char *const mapmode = map_mode_to_chars(mp->m_mode);
-      dict_T *dict = rettv->vval.v_dict;
-
-      dict_add_nr_str(dict, "lhs",     0L, lhs);
-      dict_add_nr_str(dict, "rhs",     0L, mp->m_orig_str);
-      dict_add_nr_str(dict, "noremap", mp->m_noremap ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "expr",    mp->m_expr    ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "silent",  mp->m_silent  ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "sid",     (long)mp->m_script_ID, NULL);
-      dict_add_nr_str(dict, "buffer",  (long)buffer_local, NULL);
-      dict_add_nr_str(dict, "nowait",  mp->m_nowait  ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "mode",    0L, (char_u *)mapmode);
-
-      xfree(lhs);
-      xfree(mapmode);
+      mapblock_fill_dict(rettv->vval.v_dict, mp, buffer_local, true);
     }
   }
 }
 
+/// Fill a dictionary with all applicable maparg() like dictionaries
+///
+/// @param  dict  The dictionary to be filled
+/// @param  mp  The maphash that contains the mapping information
+/// @param  buffer_value  The "buffer" value
+/// @param  compatible  True for compatible with old maparg() dict
+void mapblock_fill_dict(dict_T *const dict, const mapblock_T *const mp, long buffer_value, bool compatible)
+  FUNC_ATTR_NONNULL_ALL
+{
+  char_u *lhs = str2special_save(mp->m_keys, true);
+  char *const mapmode = map_mode_to_chars(mp->m_mode);
+
+  dict_add_nr_str(dict, "lhs",     0L, lhs);
+  dict_add_nr_str(dict, "rhs",     0L, mp->m_orig_str);
+  dict_add_nr_str(dict, "noremap", mp->m_noremap ? 1L : 0L, NULL);
```

Comment:
By the way, one thing I missed: if !compatible it should have 2 when `<script>` was used. I.e. `compatible ? !!mp->m_noremap : (mp->m_noremap == REMAP_SCRIPT ? 2 : !!mp->m_noremap)`. Currently `<script>` is not distinguishable with `noremap`, but it is essential if you want to correctly restore a mapping.